check/checksum.go

```diff
@@ -0,0 +1,182 @@
+package check
+
+import (
+	"bytes"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/schollz/progressbar/v3"
+)
+
+func isHiddenFile(f string) bool {
+	return string(f[0]) == "."
+}
+
+func checksumFor(path string) ([]byte, error) {
+	h := md5.New()
+
+	var r *os.File
+	r, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("error opening file in %s: %w", path, err)
+	}
+	defer r.Close()
+
+	if _, err := io.Copy(h, r); err != nil {
+		return nil, fmt.Errorf("error reading %s content: %w", path, err)
+	}
+
+	return h.Sum(nil), nil
+}
+
+func checkChecksum(src, target string) (bool, error) {
+	name := filepath.Base(src)
+	sc, err := checksumFor(src)
+	if err != nil {
+		return false, fmt.Errorf("error getting %s checksum: %w", src, err)
+	}
+
+	targetPath := filepath.Join(target, name)
+	tc, err := checksumFor(targetPath)
+	if err != nil {
+		return false, fmt.Errorf("error getting %s checksum: %w", targetPath, err)
+	}
+
+	return bytes.Equal(sc, tc), nil
+}
+
+// Check compares md5 checksum files from src directory with corresponding files in target directory
+func CheckChecksum(src, target string) error {
+	ls, err := filepath.Glob(filepath.Join(src, "*.md5"))
+	if err != nil {
+		return fmt.Errorf("error listing checksum files in %s directory: %w", src, err)
+	}
+
+	if len(ls) == 0 {
+		return fmt.Errorf("target directory %s has no checksum files to compare with", target)
+	}
+
+	bar := progressbar.Default(int64(len(ls)))
+	defer bar.Close()
+	bar.Describe("Checking files checksum")
+	if err := bar.RenderBlank(); err != nil {
+		return fmt.Errorf("error rendering the progress bar: %w", err)
+	}
+
+	type qm struct {
+		f     string
+		equal bool
+		err   error
+	}
+	q := make(chan qm, len(ls))
+	defer close(q)
+
+	var wg sync.WaitGroup
+
+	for _, f := range ls {
+		wg.Add(1)
+
+		go func(f string, wg *sync.WaitGroup, q chan<- qm) {
+			defer wg.Done()
+
+			equal, err := checkChecksum(f, target)
+
+			q <- qm{f, equal, err}
+		}(f, &wg, q)
+	}
+
+	wg.Wait()
+
+	var notEqual []string
+	for m := range q {
+		if m.err != nil {
+			return err
+		}
+
+		bar.Add(1)
+
+		if bar.IsFinished() {
+			if len(notEqual) > 0 {
+				return fmt.Errorf("got different checksum for file(s): %v", notEqual)
+			}
+
+			log.Output(1, "OK!")
+			return nil
+		}
+
+		if !m.equal {
+			notEqual = append(notEqual, m.f)
+		}
+	}
+
+	return nil
+}
+
+func createChecksum(src string, f string) error {
+	checksum, err := checksumFor(filepath.Join(src, f))
+	if err != nil {
+		return fmt.Errorf("error getting checksum for %s: %w", src, err)
+	}
+
+	targetPath := filepath.Join(src, fmt.Sprintf("%s.md5", f))
+	if err := os.WriteFile(targetPath, checksum, 0755); err != nil {
+		return fmt.Errorf("error writing %s checksum file: %w", targetPath, err)
```

Comment:
Não precisamos de dois argumentos aqui, né? E podemos diminuir o nome das variáveis.

```suggestion
func createChecksum(src string) error {
	checksum, err := checksumFor(src)
	if err != nil {
		return fmt.Errorf("error getting checksum for %s: %w", src, err)
	}
	pth := fmt.Sprintf("%s.md5", src)
	if err := os.WriteFile(pth, checksum, 0755); err != nil {
		return fmt.Errorf("error writing %s checksum file: %w", pth, err)
```

Reply:
Eu achei melhor passar dois argumentos porque senão vamos ter que fazer o `join` fora dela, só que eu preciso do `src` sem o arquivo para poder usar no `target`. Fazer o `join` fora, para ter que desfazer dentro não me soa muito agradável.

Reply:
O que não entendo é pq vc precisa do `target`. Voltando um pouco para o caso de uso:
* baixei os arquivos todos em `data/`
* agora quero gerar os checksum


Nesse cenário, não faz sentido em ter os arquivos baixados em `data/` e gerar o `checksum/` em outro lugar. Então o que preciso é uma função que recebe o nome de um arquivo (por exemplo, `data/Municipios.zip` e gere `data/Municipios.zip.md5`). Eu não preciso de `target/` pq:
1. o diretório onde vou criar o checksum é o mesmo do arquivo original;
2. o nome do novo arquivo é só o nome do arquivo original concateado a `.md5`.

Faz sentido?

Reply:
Ou seja:
* _input_ é `src`
* output é `fmt.Sprintf("%s.md5", src)`

Reply:
Tem razão, não tinha pensado nisso, o código estava redundante mesmo.